src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/ObstructionFreedomViolationEventsCutTest.kt

```diff
@@ -0,0 +1,81 @@
+/*
+ * Lincheck
+ *
+ * Copyright (C) 2019 - 2025 JetBrains s.r.o.
+ *
+ * This Source Code Form is subject to the terms of the
+ * Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
+ * with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package org.jetbrains.kotlinx.lincheck_test.representation
+
+import org.jetbrains.kotlinx.lincheck.annotations.Operation
+import org.jetbrains.kotlinx.lincheck.checkImpl
+import org.jetbrains.kotlinx.lincheck.isInTraceDebuggerMode
+import org.jetbrains.kotlinx.lincheck.strategy.managed.modelchecking.ModelCheckingOptions
+import org.jetbrains.kotlinx.lincheck_test.util.checkLincheckOutput
+import org.junit.Assume.assumeFalse
+import org.junit.Before
+import org.junit.Test
+import java.util.concurrent.atomic.AtomicReference
+
+/**
+ * Checks that spin-cycle repeated events are cut in case of obstruction freedom violation
+ */
+class ObstructionFreedomViolationEventsCutTest {
+
+    @Before // spin-loop detection is unsupported in trace debugger mode
+    fun setUp() = assumeFalse(isInTraceDebuggerMode)
+
+    private val q = MSQueueBlocking()
+
+    @Operation
+    fun enqueue(x: Int) = q.enqueue(x)
+
+    @Operation
+    fun dequeue(): Int? = q.dequeue()
+
+    @Test
+    fun runModelCheckingTest() = ModelCheckingOptions()
+        .checkObstructionFreedom(true)
+        .checkImpl(this::class.java)
+        .checkLincheckOutput("obstruction_freedom_violation_events_cut")
+}
+
+class MSQueueBlocking {
```

Comment:
Now implementation of `MSQueueBlocking` is added to this file, but from where was it taken from? I don't see it in the diff as removed. Is there a duplicate of class `MSQueueBlocking` right now in the source code?

Reply:
Tests from `test-integration` can include classes from `test` module. You can define `MSQueueBlocking` in `test` and just include in guide test, because there is the same class defined in `test-integration/org/jetbrains/kotlinx/lincheck_test/guide/ObstructionFreedomViolationTest.kt`